blockchain/storage/leveldb/leveldb.go

```diff
@@ -20,12 +61,105 @@ func Factory(config map[string]interface{}, logger hclog.Logger) (storage.Storag
 		return nil, fmt.Errorf("path is not a string")
 	}
 
-	return NewLevelDBStorage(pathStr, logger)
+	// get the options
+	opt := NewDefaultOptons()
+
+	cache, ok := config["cache"]
+	if ok {
+		opt.CacheSize, ok = cache.(int)
+		if !ok {
+			return nil, fmt.Errorf("cache is not a int")
+		}
+	}
+
+	handles, ok := config["handles"]
+	if ok {
+		opt.Handles, ok = handles.(int)
+		if !ok {
+			return nil, fmt.Errorf("handles is not a int")
+		}
+	}
+
+	bloomKeyBits, ok := config["bloomKeyBits"]
+	if ok {
+		opt.BloomKeyBits, ok = bloomKeyBits.(int)
+		if !ok {
+			return nil, fmt.Errorf("bloomKeyBits is not a int")
+		}
+	}
+
+	compactionTableSize, ok := config["compactionTableSize"]
+	if ok {
+		opt.CompactionTableSize, ok = compactionTableSize.(int)
+		if !ok {
+			return nil, fmt.Errorf("compactionTableSize is not a int")
+		}
+	}
+
+	compactionTotalSize, ok := config["compactionTotalSize"]
+	if ok {
+		opt.CompactionTotalSize, ok = compactionTotalSize.(int)
+		if !ok {
+			return nil, fmt.Errorf("compactionTableSize is not a int")
+		}
+	}
+
+	nosync, ok := config["nosync"]
+	if ok {
+		opt.NoSync, ok = nosync.(bool)
+		if !ok {
+			return nil, fmt.Errorf("nosync is not a bool")
+		}
+	}
+
+	return NewLevelDBStorage(pathStr, opt, logger)
 }
 
 // NewLevelDBStorage creates the new storage reference with leveldb
-func NewLevelDBStorage(path string, logger hclog.Logger) (storage.Storage, error) {
-	db, err := leveldb.OpenFile(path, nil)
+func NewLevelDBStorage(path string, o *Options, logger hclog.Logger) (storage.Storage, error) {
+	opt := &leveldbopt.Options{}
+	cache := o.CacheSize
+	handles := o.Handles
+	bloomKeyBits := o.BloomKeyBits
+	compactionTableSize := o.CompactionTableSize
+	compactionTotalSize := o.CompactionTotalSize
+
+	if o.CacheSize < minCache {
+		cache = minCache
+	}
+
+	if handles < minHandles {
+		handles = minHandles
+	}
+
+	if compactionTableSize < DefaultCompactionTableSize {
+		compactionTableSize = DefaultCompactionTableSize
+	}
+
+	if compactionTotalSize < DefaultCompactionTotalSize ||
+		compactionTotalSize < compactionTableSize {
+		compactionTotalSize = int(common.Max(
+			uint64(DefaultCompactionTotalSize),
+			uint64(compactionTableSize),
+		))
+	}
+
+	opt.OpenFilesCacheCapacity = handles
+	opt.CompactionTableSize = compactionTableSize * leveldbopt.MiB
+	opt.CompactionTotalSize = compactionTotalSize * leveldbopt.MiB
+	opt.BlockCacheCapacity = cache / 2 * leveldbopt.MiB
+	opt.WriteBuffer = cache / 4 * leveldbopt.MiB
+	opt.Filter = filter.NewBloomFilter(bloomKeyBits)
+	opt.NoSync = o.NoSync
+
+	logger.Info("leveldb", "OpenFilesCacheCapacity", opt.OpenFilesCacheCapacity)
```

Comment:
Same info here?

A better logging suggestion:

```
logger.Info("new leveldb storage", 
	"key1", val1,
	"key2", val2,
)
```

Reply:
merge log print in one line